synapse/storage/databases/main/user_directory.py

```diff
@@ -521,17 +515,39 @@ async def get_user_in_directory(self, user_id: str) -> Optional[Dict[str, Any]]:
             desc="get_user_in_directory",
         )
 
-    async def update_user_directory_stream_pos(self, stream_id: int) -> None:
+    async def update_user_directory_stream_pos(self, stream_id: Optional[int]) -> None:
         await self.db_pool.simple_update_one(
             table="user_directory_stream_pos",
             keyvalues={},
             updatevalues={"stream_id": stream_id},
             desc="update_user_directory_stream_pos",
         )
 
+    async def exclude_from_user_dir(self, user_id: str) -> bool:
+        """Certain classes of local user are omitted from the user directory.
+        Is this user one of them?
+        """
+        if self.hs.is_mine_id(user_id):
+            # Support users are for diagnostics and should not appear in the user directory.
+            if await self.is_support_user(user_id):
+                return True
+
+            # Deactivated users aren't contactable, so should not appear in the user directory.
+            if await self.get_user_deactivated_status(user_id):
+                return True
+
+            # App service users aren't usually contactable, so exclude them.
+            if self.get_if_app_services_interested_in_user(user_id):
+                return True
+
+        return False
 
-class UserDirectoryStore(UserDirectoryBackgroundUpdateStore):
 
+class UserDirectoryStore(
+    UserDirectoryBackgroundUpdateStore,
+    ApplicationServiceWorkerStore,
+    RegistrationWorkerStore,
```

Comment:
> I guess the store is one of those big mixins we were talking about [finally clicked in my head].

DataStore is a class with a lot of parents; dunno if we'd consider that a mixin. (I think of a mixin as something that's designed to be used with multiple inheritance, but isn't really a standalone Thing/Entity by itself.)

But yeah, the DataStore inherits from 9001 other Stores but... they're not all independent. Sometimes they call each other's methods.

Being honest, my motivation is
- make pycharm's ["I don't know what this method is because it's on another store"](https://user-images.githubusercontent.com/8614563/132353274-5e593c9e-5cb1-4a83-b82c-2a0c8eacd2b2.png) go away
- so that pycharm's control-click to go to definition works
- maybe make mypy happier if this ever gets mypy-ed?
- make the inter-store dependencies explicit

>  I just wonder if this is what we're meant to do. 

erm, pass. I think the status quo is "yeah it sucks, and @erikjohnston has a plan how we could do things better one day"

> I don't know enough about Python multi-inheritance rules — do we do this anywhere else? Will it cause problems?)

I had to be careful to order the parents of 

> Will it cause problems?

Was hoping to rely on unit tests to figure that out. Would appreciate input from the rest of the team here.

Having looked over it, I think this these classes get pulled in by the `UserDirectoryBackgroundUpdateStore`, not the `UserDirectoryStore`.